src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2ConnectionTests.cs

```diff
@@ -498,6 +493,21 @@ public async Task StreamPool_StreamIsInvalidState_DontReturnedToPool()
             await output._dataWriteProcessingTask.DefaultTimeout();
 
             await StopConnectionAsync(expectedLastStreamId: 1, ignoreNonGoAwayFrames: false);
+
+            async Task PingUntilStreamDisposed(Http2Stream stream)
+            {
+                var output = (Http2OutputProducer)stream.Output;
+
+                do
+                {
+                    // Ping will trigger the stream to be returned to the pool so we can assert it
+                    await SendPingAsync(Http2PingFrameFlags.NONE);
+                    await ExpectAsync(Http2FrameType.PING,
+                        withLength: 8,
+                        withFlags: (byte)Http2PingFrameFlags.ACK,
+                        withStreamId: 0);
+                } while (!output._disposed);
```

Comment:
The timeout around this method won't make this loop exit if something breaks. Do we assume that SendPingAsync or ExpectAsync will eventually throw?

Reply:
Good point.

Http2TestBase.Dispose() should get called by xunit after this test fails and will complete all transports. I believe that should make it throw.